File: historian-core/src/main/java/net/yslibrary/historian/internal/LogWritingTask.java
package net.yslibrary.historian.internal;

import net.yslibrary.historian.Historian;
import net.yslibrary.historian.LogEntity;

/**
 * Runnable implementation writing logs & executing callbacks
 */

public class LogWritingTask implements Runnable {

  private final Historian.Callbacks callbacks;
  private final LogWriter logWriter;
  private final LogEntity log;

  public LogWritingTask(Historian.Callbacks callbacks,
                        LogWriter logWriter,
                        LogEntity log) {
    this.callbacks = callbacks;
    this.logWriter = logWriter;
    this.log = log;
  }

  @Override
  public void run() {

    try {
      logWriter.log(log);

      callbacks.onSuccess();
    } catch (final Throwable throwable) {
      callbacks.onFailure(throwable);
    }
  }
}

File: historian-core/src/main/java/net/yslibrary/historian/internal/LogWritingTask.kt
package net.yslibrary.historian.internal

import net.yslibrary.historian.Historian

/**
 * Runnable implementation writing logs and executing callbacks
 */
class LogWritingTask(
    private val onSuccess: Historian.OnSuccessCallback?,
    private val onFailure: Historian.OnFailureCallback?,
    private val logWriter: LogWriter,
    private val log: LogEntity
) : Runnable {

    override fun run() {
        try {
            logWriter.log(log)
            onSuccess?.onSuccess()
        } catch (t: Throwable) {
            onFailure?.onFailure(t)
        }
    }
}
